### PR TITLE
Refactor documents

### LINKED ISSUE #303: refactor documents
refactor documents to remove code duplication

### PR DESCRIPTION
**Pull request by Bartzi**
Changes everything and hopefully makes it nicer as before
fixes #303 and obsoletes #291 

Have fun ;)

### CODE REVIEW

File: _1327/documents/templates/documents_attachments.html [hunk 1]
```diff
@@ -122,7 +128,4 @@
 
 		$('[data-toggle="tooltip"]').tooltip();
 	</script>
-
-
-{% endblock %}
-
+{% endblock %}
```
Comment (janno42) [inline _1327/documents/templates/documents_attachments.html]: newline

File: _1327/documents/forms.py [hunk 2]
```diff
@@ -37,42 +38,93 @@ def clean_url_title(self):
...
+		# make sure that view permission is enabled if other permissions are enabled
+		view_permission_enabled = False
+		other_permission_enabled = False
+		cleaned_data = super(PermissionBaseForm, self).clean()
+		for field, __ in filter(lambda x: type(x[1]) == forms.BooleanField, self.fields.items()):
+			field_value = cleaned_data.get(field)
+			if 'view' in field:
+				view_permission_enabled = field_value
+			else:
+				if not other_permission_enabled:
```
Comment (janno42) [inline _1327/documents/forms.py]: `other_permission_enabled = other_permission_enabled or field_value`

File: _1327/documents/forms.py [hunk 3]
```diff
@@ -37,42 +38,93 @@ def clean_url_title(self):
...
+					other_permission_enabled = field_value
+		if other_permission_enabled and not view_permission_enabled:
+			raise forms.ValidationError("If you want to enable additional permissions for a group you also need to enable the view permission for that group!")
```
Comment (janno42) [inline _1327/documents/forms.py]: translate!

File: _1327/documents/templates/documents_edit.html [hunk 4]
```diff
@@ -1,16 +1,19 @@
...
+	{% endif %}
 
-{% for new_autosaved_page in new_autosaved_pages %}
-	<div class="alert alert-danger alert-margin-bottom" role="alert">{% blocktrans with date=new_autosaved_page.created %}There is a new document autosaved on {{ date }}{% endblocktrans %}, {% trans "you can restore this unsaved page:" %} <btn href="{{ new_autosaved_page.get_edit_url }}?restore" class="btn-sm btn-default">{% trans "Restore" %}</a></div>
-{%  endfor %}
+	{% for new_autosaved_page in new_autosaved_pages %}
+		<div class="alert alert-danger alert-margin-bottom" role="alert">{% blocktrans with date=new_autosaved_page.created %}There is a new document autosaved on {{ date }}{% endblocktrans %}, {% trans "you can restore this unsaved page:" %} <btn href="{{ new_autosaved_page.get_edit_url }}?restore" class="btn-sm btn-default">{% trans "Restore" %}</a></div>
+	{%  endfor %}
 
 <form action="{{ document.get_edit_url }}" method="post" class="form-horizontal" role="form" id="document-form">
```
Comment (janno42) [inline _1327/documents/templates/documents_edit.html]: indentation

File: _1327/information_pages/templates/information_pages_meta_information.html [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{% load i18n %}
+
+<dl class="meta-information">
+    <dt>{% blocktrans count counter=document.authors|length %}Author{% plural %}Authors{% endblocktrans %}</dt>
+    <dd>{% for author in document.authors %}{{ author }}{% if not forloop.last %}, {% endif %}{% endfor %}</dd>
+    <dt>{% trans "Created" %}</dt>
+    <dd>{{ document.created|date:"d. M Y, H:i" }}</dd>
+    <dt>{% trans "Last change" %}</dt>
+	<dd>{{ document.last_change|date:"d. M Y, H:i" }}</dd>
+</dl>
```
Comment (janno42) [inline _1327/information_pages/templates/information_pages_meta_information.html]: newline

File: _1327/main/migrations/0003_menu_item_data.py [hunk 6]
```diff
@@ -36,15 +38,16 @@ def initial_data(apps, schema_editor):
...
 	dependencies = [
-		('main', '0007_menuitem_menu_type'),
+		('main', '0002_add_staff_group'),
 	]
 
+
 	operations = [
 		migrations.RunPython(initial_data, delete_menu_data),
-	]
+	]
```
Comment (janno42) [inline _1327/main/migrations/0003_menu_item_data.py]: newline

File: _1327/minutes/templates/minutes_edit.html [hunk 7]
```diff
@@ -1,5 +1,49 @@
...
+    $('.delete-button').each(function() {
+        $(this).prev(".input-group.inlineformset").children(".input-group-btn").append(this);
+    });
+    $('.add-row-container').append($('.add-row'));
+    $('.add-row').addClass('btn btn-default');
+    $('.add-row').text('{% trans "Add another guest" %}');
+
+    // add datepicker to date inputs
+    $("input[name*='date']:not([readonly='True'])").datepicker( $.datepicker.regional["{{ LANGUAGE_CODE }}"] );
+{% endblock %}
```
Comment (janno42) [inline _1327/minutes/templates/minutes_edit.html]: newline

File: _1327/minutes/templates/minutes_list.html [hunk 8]
```diff
@@ -39,4 +39,4 @@ <h2 id="year{{ year.year }}">{{ year.year }}</h2>
 			{% endfor %}
 		</table>
 	{% endfor %}
-{% endblock %}
+{% endblock %}
```
Comment (janno42) [inline _1327/minutes/templates/minutes_list.html]: newline

File: _1327/minutes/templates/minutes_meta_information.html [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    {% endfor %}
+    <dt>{% trans "Guests" %}</dt>
+    {% for guest in document.guests.all %}
+    <dd>{{ guest.name }}</dd>
+    {% empty %}
+    <dd>{% trans "No Guests" %}</dd>
+    {% endfor %}
+    <dt>{% trans "Last change" %}</dt>
+    <dd>{{ document.last_change|date:"d. M Y, H:i" }}</dd>
+</dl>
```
Comment (janno42) [inline _1327/minutes/templates/minutes_meta_information.html]: newline

File: _1327/minutes/templates/minutes_meta_information.html [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+<dl class="meta-information">
+    <dt>{% trans "Date" %}</dt>
+    <dd>{{ document.date|date:"d. M Y" }}</dd>
+    <dt>{% trans "Moderator" %}</dt>
+    <dd>{{ document.moderator }}</dd>
+    <dt>{% trans "Minute taker" %}</dt>
+    <dd>{{ document.author }}</dd>
+    <dt>{% trans "Participants" %}</dt>
+    {% for participant in document.participants.all %}
+    <dd>{{ participant }}</dd>
```
Comment (janno42) [inline _1327/minutes/templates/minutes_meta_information.html]: indentation here and below

File: _1327/polls/templates/polls_create_poll.html [hunk 11]
```diff
@@ -39,6 +39,19 @@
 				{% endfor %}
 			</tbody>
 		</table>
+        <h3>{% trans "Permissions" %}</h3>
+        <div class="divider"></div>
+        {{ permission_formset.management_form }}
+        <table class="table table-striped vertically-aligned table-narrow" >
```
Comment (janno42) [inline _1327/polls/templates/polls_create_poll.html]: space before closing `>` and indentations in the table below

**Comment by janno42** [comment]
it's getting pretty nice ;)

File: _1327/documents/templatetags/document_perms.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+
+class ObjectPermissionsNode(GuardianObjectPermissionChecker):
+
+	def __init__(self, for_whom, obj, context_var):
+		super(ObjectPermissionsNode, self).__init__(str(for_whom), str(obj), str(context_var))
+
+	def render(self, context):
+		super(ObjectPermissionsNode, self).render(context)
+		context[self.context_var] = list(map(lambda x: x.split('_')[0], context[self.context_var]))
```
Comment (invliD) [inline _1327/documents/templatetags/document_perms.py]: Documentation!

File: _1327/documents/models.py [hunk 13]
```diff
@@ -55,6 +55,10 @@ def authors(self):
 		return authors
 
 	@property
+	def meta_information_text(self):
```
Comment (invliD) [inline _1327/documents/models.py]: Please rename. `text` implies no HTML.

File: _1327/information_pages/__init__.py [hunk 14]
```diff
@@ -0,0 +1 @@
+from .forms import InformationDocumentForm  # noqa
```
Comment (invliD) [inline _1327/information_pages/__init__.py]: Doesn't this introduce the deprecation warnings again?

**Comment by invliD** [comment]
:+1:

**Pull request merged**